x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/process/IndexingStateProcessorTests.java

```diff
@@ -56,54 +64,92 @@
+        verifyNoMoreInteractions(resultsPersisterService);
     }
 
-    public void testStateRead() throws IOException {
+    public void testExtractDocId() throws IOException {
+        assertThat(IndexingStateProcessor.extractDocId(new BytesArray(STATE_SAMPLE.getBytes(StandardCharsets.UTF_8))), equalTo("1"));
+    }
+
+    private void testStateRead(SearchHits searchHits, String expectedIndexOrAlias) throws IOException {
+        when(searchResponse.getHits()).thenReturn(searchHits);
+
         ByteArrayInputStream stream = new ByteArrayInputStream(STATE_SAMPLE.getBytes(StandardCharsets.UTF_8));
         stateProcessor.process(stream);
         ArgumentCaptor<BytesReference> bytesRefCaptor = ArgumentCaptor.forClass(BytesReference.class);
-        verify(stateProcessor, times(3)).persist(bytesRefCaptor.capture());
+        verify(stateProcessor, times(3)).persist(eq(expectedIndexOrAlias), bytesRefCaptor.capture());
 
         String[] threeStates = STATE_SAMPLE.split("\0");
         List<BytesReference> capturedBytes = bytesRefCaptor.getAllValues();
         assertEquals(threeStates[0], capturedBytes.get(0).utf8ToString());
         assertEquals(threeStates[1], capturedBytes.get(1).utf8ToString());
         assertEquals(threeStates[2], capturedBytes.get(2).utf8ToString());
+        verify(resultsPersisterService, times(3)).searchWithRetry(any(SearchRequest.class), any(), any(), any());
         verify(resultsPersisterService, times(3)).bulkIndexWithRetry(any(BulkRequest.class), any(), any(), any());
     }
 
+    public void testStateRead_StateDocumentCreated() throws IOException {
+        testStateRead(SearchHits.empty(), ".ml-state-write");
+    }
+
+    public void testStateRead_StateDocumentUpdated() throws IOException {
+        testStateRead(
+            new SearchHits(new SearchHit[]{ SearchHit.createFromMap(Map.of("_index", ".ml-state-dummy")) }, null, 0.0f),
+            ".ml-state-dummy");
+    }
+
     public void testStateReadGivenConsecutiveZeroBytes() throws IOException {
         String zeroBytes = "\0\0\0\0\0\0";
         ByteArrayInputStream stream = new ByteArrayInputStream(zeroBytes.getBytes(StandardCharsets.UTF_8));
 
         stateProcessor.process(stream);
 
-        verify(stateProcessor, never()).persist(any());
-        Mockito.verifyNoMoreInteractions(resultsPersisterService);
+        verify(stateProcessor, never()).persist(any(), any());
     }
 
     public void testStateReadGivenConsecutiveSpacesFollowedByZeroByte() throws IOException {
-        String zeroBytes = "        \n\0";
-        ByteArrayInputStream stream = new ByteArrayInputStream(zeroBytes.getBytes(StandardCharsets.UTF_8));
+        String bytes = "        \n\0";
+        ByteArrayInputStream stream = new ByteArrayInputStream(bytes.getBytes(StandardCharsets.UTF_8));
 
-        stateProcessor.process(stream);
+        Exception e = expectThrows(IllegalStateException.class, () -> stateProcessor.process(stream));
```

Comment:
This looks like a dangerous change.  Previously if the C++ wrote nothing but whitespace it was silently ignored.  Now it will cause the state processor to throw an exception and never process any subsequent state.

I think the behaviour should be changed back to what it was for the pure whitespace case.  It's fine to throw an exception if a bulk metadata JSON object is invalid.  But if it's not present at all then I think we should maintain the previous behaviour of treating it as a no-op.  Otherwise a very careful audit of the C++ code will be required to find what situations it might write blank state.

Reply:
Good finding, done.
Please note that I'm now skipping leading blank lines just in case the actual request body is present *after* those blank lines.